metrics/internal/collectors/storage-auto-scaler.go

```diff
@@ -0,0 +1,174 @@
+	ds := []*prometheus.Desc{
+		s.StorageAutoScalerMetadata,
+		s.AutoScalerPhaseStatus,
+		s.StorageCapacityReached,
+	}
+	for _, d := range ds {
+		ch <- d
+	}
+}
+
+func (s *StorageAutoScalerCollector) Collect(ch chan<- prometheus.Metric) {
+	sasLister := NewStorageAutoScalerLister(s.Informer.GetIndexer())
+	storageAutoScalers, err := sasLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("couldn't list StorageAutoScaler elements: %v", err)
+		return
+	} else if len(storageAutoScalers) == 0 {
+		return
+	}
+	var wg sync.WaitGroup
+	wg.Add(3) // we are waiting for THREE go routines to finish
+	go s.collectStorageAutoScalerMetadata(ch, storageAutoScalers, &wg)
+	go s.collectStorageAutoScalerPhaseStatuses(ch, storageAutoScalers, &wg)
+	go s.collectStorageCapacityReachedStatuses(ch, storageAutoScalers, &wg)
+	wg.Wait()
+}
+
+func (s *StorageAutoScalerCollector) Run(stopCh <-chan struct{}) {
+	go s.Informer.Run(stopCh)
+}
+
+func (s *StorageAutoScalerCollector) collectStorageAutoScalerMetadata(ch chan<- prometheus.Metric, storageAutoScalers []*v1.StorageAutoScaler, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for _, storageAutoScaler := range storageAutoScalers {
+		ch <- prometheus.MustNewConstMetric(
+			s.StorageAutoScalerMetadata,
+			prometheus.GaugeValue,
+			float64(1),
+			storageAutoScaler.Name,
+			storageAutoScaler.Namespace,
+			storageAutoScaler.Spec.StorageCluster.Name,
+			storageAutoScaler.Spec.DeviceClass,
+		)
+	}
+}
+
+func (s *StorageAutoScalerCollector) collectStorageAutoScalerPhaseStatuses(ch chan<- prometheus.Metric, storageAutoScalers []*v1.StorageAutoScaler, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for _, storageAutoScaler := range storageAutoScalers {
+		var phaseStatus = NotStartedPhaseStatus
+		switch storageAutoScaler.Status.Phase {
+		case v1.StorageAutoScalerPhaseFailed:
+			phaseStatus = FailedPhaseStatus
+		case v1.StorageAutoScalerPhaseSucceeded:
+			phaseStatus = succeededPhaseValidation(storageAutoScaler.Status.LastExpansion.CompletionTime)
+		case v1.StorageAutoScalerPhaseInProgress:
+			phaseStatus = InProgressPhaseStatus
+		default:
+			phaseStatus = NotStartedPhaseStatus
+		}
+		// send metric on any phase other than 'Reset' phase
+		if phaseStatus != ResetPhaseStatus {
+			ch <- prometheus.MustNewConstMetric(
+				s.AutoScalerPhaseStatus,
+				prometheus.GaugeValue,
+				float64(phaseStatus),
+				storageAutoScaler.Name,
+				storageAutoScaler.Namespace,
+			)
+		}
+	}
+}
+
+func succeededPhaseValidation(completionTime metav1.Time) PhaseStatus {
+	currentTime := metav1.Now().Time
+	var zeroValuedTime metav1.Time
+	// if completion time is not set
+	// OR
+	// success happened within a 10mins time window
+	// we return success
+	if completionTime.Sub(zeroValuedTime.Time) == 0 ||
+		currentTime.Sub(completionTime.Time) < 10*time.Minute {
+		return SucceededPhaseStatus
+	}
+	// after 10 mins, we will reset the success phase
+	return ResetPhaseStatus
+}
```

Comment:
@aruniiird @parth-gr I still feel this is the job of the operator to define the status of the CR. Current implementation causes a discrepancy between `storageAutoScaler.Status.Phase` values and these metrics.

My suggestion: Reconcile StorageAutoScalerCR. If it has been 10minutes since the last `v1.StorageAutoScalerPhaseSucceeded`, change the status to `NotStarted`. Then the metrics will automatically change and alert will be stopped.

Reply:
Controller should never change the state to `NotStarted`, it is only the state when there is no reconcile started.

As this logic if purely needed for alert we would like to keep it here as discussed.

Later if we can have toast notification we can even remove this

Reply:
This what decided for 4.19, we can plan improvement once UI is unblock of using toast notif.

Reply:
Sure raise it for 10mins, but I do not agree to the current implementation and consider it as a blocker.
To be clear, I do not have any issue on rest of the implementation, just this function that creates a hypothetical state which does not really exist.

Reply:
On second thought, since the state is mostly internal and we do not expose it as a metric, it's fine to merge as-is.
Refactors to it does not affect user outcome.

Reply:
To be clear `ResetPhaseStatus ` dont reset the actual CR state, it just for the alert to understand when to stop sending it, correct ?

Reply:
Absolutely correct, `ResetPhaseStatus` is an internal phase, which determines when to send the metrics (to prometheus) or not and it NEVER appears as an actual user visible phase.